selfdrive/hardware/tici/agnos.py

```diff
@@ -74,25 +76,75 @@ def unsparsify(f: StreamingDecompressor) -> Generator[bytes, None, None]:
       raise Exception("Unhandled sparse chunk type")
 
 
-def flash_partition(cloudlog, spinner, target_slot, partition):
-  cloudlog.info(f"Downloading and writing {partition['name']}")
+def get_target_slot_number():
+  current_slot = subprocess.check_output(["abctl", "--boot_slot"], encoding='utf-8').strip()
+  return 1 if current_slot == "_a" else 0
 
-  downloader = StreamingDecompressor(partition['url'])
-  with open(f"/dev/disk/by-partlabel/{partition['name']}{target_slot}", 'wb+') as out:
-    partition_size = partition['size']
 
-    # Check if partition is already flashed
-    out.seek(partition_size)
-    if out.read(64) == partition['hash_raw'].lower().encode():
-      cloudlog.info(f"Already flashed {partition['name']}")
-      return
+def slot_number_to_suffix(slot_number):
+  assert slot_number in (0, 1)
+  return '_a' if slot_number == 0 else '_b'
+
+
+def get_partition_path(target_slot_number, partition):
+  path = f"/dev/disk/by-partlabel/{partition['name']}"
+
+  if partition.get('has_ab', True):
+    path += slot_number_to_suffix(target_slot_number)
+
+  return path
+
+
+def verify_partition(target_slot_number, partition):
+  full_check = partition.get('full_check', False)
+  path = get_partition_path(target_slot_number, partition)
+  partition_size = partition['size']
+
+  with open(path, 'rb+') as out:
+    if full_check:
+      raw_hash = hashlib.sha256()
+
+      pos = 0
+      chunk_size = 1024 * 1024
+      while pos < partition_size:
+        n = min(chunk_size, partition_size - pos)
+        raw_hash.update(out.read(n))
+        pos += n
+
+      return raw_hash.hexdigest().lower() == partition['hash_raw'].lower()
+    else:
+      out.seek(partition_size)
+      return out.read(64) == partition['hash_raw'].lower().encode()
+
+
+def clear_partition_hash(target_slot_number, partition):
+  path = get_partition_path(target_slot_number, partition)
+  with open(path, 'wb+') as out:
+    partition_size = partition['size']
 
-    # Clear hash before flashing
     out.seek(partition_size)
     out.write(b"\x00" * 64)
-    out.seek(0)
     os.sync()
 
+
+def flash_partition(target_slot_number, partition, cloudlog, spinner=None):
+  cloudlog.info(f"Downloading and writing {partition['name']}")
+
+  if verify_partition(target_slot_number, partition):
+    cloudlog.info(f"Already flashed {partition['name']}")
+    return
+
+  downloader = StreamingDecompressor(partition['url'])
+
+  # Clear hash before flashing in case we get interrupted
+  full_check = partition.get('full_check', False)
```

Comment:
safer to default to true
```suggestion
  full_check = partition.get('full_check', True)
```

Reply:
Was for backwards compatibility with the old manifest. But that's updated now.